integrations/tensorflow/e2e/iree_tf_benchmark_suite.cmake

```diff
@@ -0,0 +1,112 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# iree_tf_benchmark_suite()
+#
+# Generates benchmark suites for TensorFlow Python models/layers.
+#
+# Note that this CMake function works closely with tf_benchmark_utils.py;
+# it uses command-line options in that script.
+#
+# Parameters:
+#   MODELS: A list of models to generate benchmark artifacts for.
+#   MODEL_SCRIPT: The Python script used to generate benchmark artifacts.
+#   MODEL_SCRIPT_ARGS: A list of command-line options and their values to the
+#       model script.
+#   NEED_ARG_FOR_MODEL: Whether an additional "--model=<model>" command-line
+#       option should be given to the model generation script.
+#   TARGET_BACKEND: The target backend to generate benchmark artifacts for.
+#   TARGET_ARCH: The detailed target backend's architecture.
+#   COMPILATION_FLAGS: A list of command-line options and their values to
+#       pass to the compiler for artifact generation.
+#   RUNTIME_FLAGS: A list of command-line options and their values to pass
+#       to the runtime when the benchmark is invoked.
+#   BENCHMARK_MODE: The mode of this benchmark suite. This does not need
+#       to be a unique identifier for this suite, which will also consider
+#       the model, target backend, and architecture. It is mainly used to
+#       differentiate suites on top of that.
+#
+function(iree_tf_benchmark_suite)
+  if(NOT IREE_BUILD_BENCHMARKS)
+    return()
+  endif()
+
+  cmake_parse_arguments(
+    PARSE_ARGV 0
+    _RULE
+    "NEED_ARG_FOR_MODEL"
+    "BENCHMARK_MODE;MODEL_SCRIPT;TARGET_BACKEND;TARGET_ARCH"
+    "COMPILATION_FLAGS;MODELS;MODEL_SCRIPT_ARGS;RUNTIME_FLAGS"
+  )
+
+  # Generate all benchmarks to the root build directory. This helps for
+  # discovering them and launch them on devices.
+  set(_ARTIFACTS_DIR "${CMAKE_BINARY_DIR}/benchmark_suites/tf_models")
+
+  foreach(_MODEL IN LISTS _RULE_MODELS)
+    # Construct the benchmark generation target name, which is the model name
+    # followed by target backend and configuration.
+    set(_NAME_LIST "generate_benchmark_artifact")
+    list(APPEND _NAME_LIST "${_MODEL}")
+    list(APPEND _NAME_LIST "${_RULE_BENCHMARK_MODE}")
+    list(APPEND _NAME_LIST "${_RULE_TARGET_BACKEND}")
+    list(APPEND _NAME_LIST "${_RULE_TARGET_ARCH}")
+    list(JOIN _NAME_LIST "__" _NAME)
+
+    # Add a command-line option to specify the model if needed.
+    set(_MODEL_CL "")
+    if(_RULE_NEED_ARG_FOR_MODEL)
+      set(_MODEL_CL "--model=${_MODEL}")
+    endif()
+
+    # Add a command-line option to specify the runtime flags if specified.
+    set(_RUNTIME_FLAGS_CL "")
+    if(_RULE_RUNTIME_FLAGS)
+      set(_RUNTIME_FLAGS_CL "--runtime_flags=\"${_RULE_RUNTIME_FLAGS}\"")
+    endif()
+
+    set(_COMBINED_CONFG_NAME "${_RULE_TARGET_ARCH}__${_RULE_BENCHMARK_MODE}")
+
+    add_custom_target("${_NAME}"
+      COMMAND
```

Comment:
The command for `add_custom_target` is *always considered out of date* (https://cmake.org/cmake/help/latest/command/add_custom_target.html). I think we want `add_custom_command` wrapped in a custom target instead. This is not something we want to execute unnecessarily

Reply:
Actually I think it's fine to use `add_custom_target`. Note that this is meant to be invoked *manually* and very infrequently. (It's not included in `ALL` even if with `-DIREE_BUILD_BENCHMARKS=ON`.)